_posts/2020-12-01-the-road-to-the-jit.md

```diff
@@ -0,0 +1,382 @@
+
+Why is that?
+
+The main reason is probably that many typical huge Erlang applications
+don't contain enough sequential code that HiPE could optimize.  The
+runtime of those systems is typically dominated by some combination of
+message passing, calls to the [ETS BIFs][ets], and garbage collection, none
+of which HiPE can optimize.
+
+[ets]: http://erlang.org/doc/man/ets.html
+
+Another reason could be that many systems have many small modules.  The
+HiPE native compiler (in common with the Erlang compiler) cannot optimize
+code across module boundaries, thus being unable to do much type-based
+optimizations.
+
+Also for most systems, it was not possible to compile all code to
+native code.  There is a small cost of switching from native code to
+interpreted BEAM and vice versa.  It is a non-trivial task to figure
+out which modules that would gain from being compiled to native code, and
+at the same time avoiding an excessive amount of context switches between
+native and interpreted code.
+
+Because none of the Ericsson Erlang projects used the HiPE native
+compiler, the OTP team could only afford to spend a limited amount of
+time maintaining HiPE. Therefore, the documentation for HiPE includes
+this note:
+
+> HiPE and execution of HiPE compiled code only have limited support
+  by the OTP team at Ericsson. The OTP team only does limited
+  maintenance of HiPE and does not actively develop HiPE. HiPE is
+  mainly supported by the HiPE team at Uppsala University.
+
+### Other Outcomes the HiPE Project
+
+I think it is fair to say that Erlang/OTP would look very different
+today if it hasn't been for the HiPE project.  Here are the major
+contributions from the HiPE project to OTP:
+
+* A new [staged tag scheme][newtags] in OTP R7.  The new tag scheme
+  allowed the Erlang system to address the full 4GB address space (the
+  previous tag scheme only supported addressing the lower 1 GB).
+  Surprisingly, the new tag scheme also improved performance.
+
+* The [Core Erlang][core] intermediate representation used in the
+  Erlang compiler to this day.  For more information, see
+  [An introduction to Core Erlang][core_introduction] and
+  [Core Erlang by Example][core_by_example].
+
+* [Dialyzer][dialyzer] (DIscrepancy AnaLYZer for ERlang programs),
+  started out as a type analysis pass for the HiPE native compiler,
+  but soon become a tool for Erlang programmers to help find bugs and
+  unreachable code in their applications.
+
+* [Bit strings and binary comprehensions][bitstrings].
+
+* Introducing [`try`...`catch`][exception] in OTP R10.
+
+* Repeatedly suggesting that Erlang needed a literal pool for premade
+  literal terms (instead of constructing them each time they are
+  used).  At one of our meetings between the HiPE team and the OTP
+  team, I remember [Richard Carlsson][richcarl] pointing out to me that it would
+  be nice for [Wings3D][wings3d] to have floating point literals.  The
+  OTP team implemented literal pools in OTP R12.
```

Comment:
It is not clear to me if floating point literals have any connection to literal pools.

Reply:
Floating point constants are stored in the literal pool.